In [2]:
import pandas as pd
import math
import numpy as np

In [15]:
df = pd.read_csv("Flights_mergedSUM_Degrees.csv")
df.head(50)

Id  Label                                          name  \
0    1    NaN                                Goroka Airport   
1    2    NaN                                Madang Airport   
2    3    NaN                  Mount Hagen Kagamuga Airport   
3    4    NaN                                Nadzab Airport   
4    5    NaN   Port Moresby Jacksons International Airport   
5    6    NaN                   Wewak International Airport   
6    7    NaN                            Narsarsuaq Airport   
7    8    NaN                       Godthaab / Nuuk Airport   
8    9    NaN                         Kangerlussuaq Airport   
9   10    NaN                                Thule Air Base   
10  11    NaN                              Akureyri Airport   
11  12    NaN                           Egilsstaðir Airport   
12  15    NaN                            Ísafjörður Airport   
13  16    NaN                Keflavik International Airport   
14  18    NaN                             Reykjavik Airport   
15  21    NaN                       Sault Ste Marie Airport   
16  24    NaN                           St. Anthony Airport   
17  27    NaN                           Baie Comeau Airport   
18  28    NaN                                CFB Bagotville   
19  29    NaN                            Baker Lake Airport   
20  30    NaN                        Campbell River Airport   
21  31    NaN                     Brandon Municipal Airport   
22  32    NaN                         Cambridge Bay Airport   
23  33    NaN                               Nanaimo Airport   
24  34    NaN      Castlegar/West Kootenay Regional Airport   
25  36    NaN                                Charlo Airport   
26  37    NaN                             Kugluktuk Airport   
27  40    NaN                           Clyde River Airport   
28  41    NaN                         Coral Harbour Airport   
29  45    NaN                             Deer Lake Airport   
30  48    NaN                          Dawson Creek Airport   
31  49    NaN                Edmonton International Airport   
32  50    NaN                                Arviat Airport   
33  54    NaN                     Inuvik Mike Zubko Airport   
34  55    NaN                               Iqaluit Airport   
35  56    NaN                           Fredericton Airport   
36  58    NaN                             Flin Flon Airport   
37  60    NaN                          Fort Simpson Airport   
38  61    NaN                Kingston Norman Rogers Airport   
39  62    NaN                     La Grande Rivière Airport   
40  63    NaN                Gaspé (Michel-Pouliot) Airport   
41  65    NaN                  Îles-de-la-Madeleine Airport   
42  67    NaN                       Dryden Regional Airport   
43  68    NaN                     Ulukhaktok Holman Airport   
44  69    NaN                            Gjoa Haven Airport   
45  70    NaN  John C. Munro Hamilton International Airport   
46  71    NaN               Montréal / Saint-Hubert Airport   
47  72    NaN             Hay River / Merlyn Carter Airport   
48  73    NaN     Halifax / Stanfield International Airport   
49  75    NaN                            Pond Inlet Airport   

                    city           country iata  icao   latitude   longitude  \
0                 Goroka  Papua New Guinea  GKA  AYGA  -6.081690  145.391998   
1                 Madang  Papua New Guinea  MAG  AYMD  -5.207080  145.789001   
2            Mount Hagen  Papua New Guinea  HGU  AYMH  -5.826790  144.296005   
3                 Nadzab  Papua New Guinea  LAE  AYNZ  -6.569803  146.725977   
4           Port Moresby  Papua New Guinea  POM  AYPY  -9.443380  147.220001   
5                  Wewak  Papua New Guinea  WWK  AYWK  -3.583830  143.669006   
6           Narssarssuaq         Greenland  UAK  BGBW  61.160500  -45.425999   
7               Godthaab         Greenland  GOH  BGGH  64.190903  -51.678101   
8            Sondrestrom         Greenland  SFJ  BGSF  67.012222  -50.

In [4]:
df.describe()

Id  Label     latitude    longitude           alt  \
count   3147.000000    0.0  3147.000000  3147.000000   3147.000000   
mean    3576.896409    NaN    25.817445     3.425252    996.084207   
std     2315.779158    NaN    27.764499    90.669891   1740.382501   
min        1.000000    NaN   -54.843300  -179.876999    -72.000000   
25%     1757.000000    NaN     6.942539   -77.197452     48.000000   
50%     3271.000000    NaN    32.898102     9.604000    262.000000   
75%     5670.500000    NaN    46.177900    79.901352   1083.000000   
max    11922.000000    NaN    78.246101   179.341003  14472.000000   

          indegree    outdegree       Degree  weighted indegree  \
count  3147.000000  3147.000000  3147.000000        3147.000000   
mean     11.698761    11.698761    23.397521          21.188116   
std      24.908456    24.980071    49.883430          55.179337   
min       1.000000     1.000000     2.000000           1.000000   
25%       2.000000     2.000000     4.000000           2.000000   
50%       3.000000     3.000000     6.000000           4.000000   
75%       9.000000     9.000000    18.000000          13.000000   
max     238.000000   239.000000   477.000000         911.000000   

       weighted outdegree  Weighted Degree  Eccentricity  closnesscentrality  \
count         3147.000000      3147.000000   3147.000000         3147.000000   
mean            21.188116        42.376231      8.656498            0.259067   
std             55.264873       110.437920      0.791659            0.041494   
min              1.000000         2.000000      7.000000            0.117349   
25%              2.000000         4.000000      8.000000            0.234077   
50%              4.000000         8.000000      9.000000            0.260452   
75%             13.000000        26.000000      9.000000            0.286365   
max            915.000000      1826.000000     13.000000            0.414766   

       harmonicclosnesscentrality  betweenesscentrality  
count                 3147.000000           3147.000000  
mean                     0.274919           9340.054020  
std                      0.048845          43132.734951  
min                      0.118753              0.000000  
25%                      0.244445              0.000000  
50%                      0.274720              1.656687  
75%                      0.305267           1112.051270  
max                      0.466609         634112.601426

On average, an airport has how many outgoing flights? -> average weighted outdegree: 21.188
And to how many diferent airports? -> average outdegree: 11.699

In [5]:
df["outdegree"].mean()

np.float64(11.698760724499524)

In [6]:
df[df["Id"] == 3830]


Id  Label                                  name     city  \
1867  3830    NaN  Chicago O'Hare International Airport  Chicago   

            country iata  icao  latitude  longitude  alt  ... indegree  \
1867  United States  ORD  KORD   41.9786   -87.9048  672  ...      203   

     outdegree Degree  weighted indegree  weighted outdegree  Weighted Degree  \
1867       205    408                550                 557             1107   

      Eccentricity  closnesscentrality  harmonicclosnesscentrality  \
1867             7            0.388491                    0.436327   

      betweenesscentrality  
1867         457824.881245  

[1 rows x 23 columns]

In [7]:
df[df["Id"] == 3682]

Id  Label                                              name     city  \
1792  3682    NaN  Hartsfield Jackson Atlanta International Airport  Atlanta   

            country iata  icao  latitude  longitude   alt  ... indegree  \
1792  United States  ATL  KATL   33.6367 -84.428101  1026  ...      216   

     outdegree Degree  weighted indegree  weighted outdegree  Weighted Degree  \
1792       217    433                911                 915             1826   

      Eccentricity  closnesscentrality  harmonicclosnesscentrality  \
1792             7            0.375911                    0.427594   

      betweenesscentrality  
1792         306637.099759  

[1 rows x 23 columns]

In [8]:
df_edges = pd.read_csv("Flights_mergedSUM_Edges.csv")
df_edges.head()

Source  Target      Type      Id  Label  Weight airline   aid    s    d  \
0    2965    2990  Directed  270652    NaN       1      2B   410  AER  KZN   
1    2966    2990  Directed  270653    NaN       1      2B   410  ASF  KZN   
2    2966    2962  Directed  270654    NaN       1      2B   410  ASF  MRV   
3    2968    2990  Directed  270655    NaN       1      2B   410  CEK  KZN   
4    2968    4078  Directed  270656    NaN       2      2G  1654  CEK  OVB   

  codeshare  stops equipment  
0       NaN      0       CR2  
1       NaN      0       CR2  
2       NaN      0       CR2  
3       NaN      0       CR2  
4       NaN      0       A81

In [9]:
df_edges_sorted = df_edges.sort_values(by="Weight", ascending=False)
df_edges_sorted.head()


Source  Target      Type      Id  Label  Weight airline   aid    s    d  \
5558    3830    3682  Directed  276882    NaN      20      VS  5347  ORD  ATL   
4066    3682    3830  Directed  275354    NaN      19      VS  5347  ATL  ORD   
2685    3179    3885  Directed  273623    NaN      13      TG  4940  HKT  BKK   
5631    3830    3861  Directed  276955    NaN      13      US  5265  ORD  MSY   
4065    3682    3576  Directed  275353    NaN      12      VS  5347  ATL  MIA   

     codeshare  stops                    equipment  
5558         Y      0                  717 M88 319  
4066         Y      0      757 319 320 M90 M88 717  
2685         Y      0  333 744 330 AB6 773 747 320  
5631         Y      0                          E75  
4065         Y      0                      757 320

In [10]:
df_edges_sorted["Source_name"] = df_edges_sorted["Source"].map(df.set_index("Id")["name"])
df_edges_sorted


Source  Target      Type      Id  Label  Weight airline   aid    s  \
5558     3830    3682  Directed  276882    NaN      20      VS  5347  ORD   
4066     3682    3830  Directed  275354    NaN      19      VS  5347  ATL   
2685     3179    3885  Directed  273623    NaN      13      TG  4940  HKT   
5631     3830    3861  Directed  276955    NaN      13      US  5265  ORD   
4065     3682    3576  Directed  275353    NaN      12      VS  5347  ATL   
...       ...     ...       ...     ...    ...     ...     ...   ...  ...   
36811    3330    6337  Directed  338305    NaN       1      ZL  4178  TSV   
36812    3363    3339  Directed  338306    NaN       1      ZL  4178  WGA   
36813    6337    6289  Directed  338308    NaN       1      ZL  4178  WIN   
36814    6337    3330  Directed  338309    NaN       1      ZL  4178  WIN   
36815    6334    3341  Directed  338310    NaN       1      ZL  4178  WYA   

         d codeshare  stops                    equipment  \
5558   ATL         Y      0                  717 M88 319   
4066   ORD         Y      0      757 319 320 M90 M88 717   
2685   BKK         Y      0  333 744 330 AB6 773 747 320   
5631   MSY         Y      0                          E75   
4065   MIA         Y      0                      757 320   
...    ...       ...    ...                          ...   
36811  WIN       NaN      0                          SF3   
36812  MEL       NaN      0                          SF3   
36813  LRE       NaN      0                          SF3   
36814  TSV       NaN      0                          SF3   
36815  ADL       NaN      0                          SF3   

                                            Source_name  
5558               Chicago O'Hare International Airport  
4066   Hartsfield Jackson Atlanta International Airport  
2685                       Phuket International Airport  
5631               Chicago O'Hare International Airport  
4065   Hartsfield Jackson Atlanta International Airport  
...                                                 ...  
36811                                Townsville Airport  
36812                          Wagga Wagga City Airport  
36813                                    Winton Airport  
36814                                    Winton Airport  
36815                                   Whyalla Airport  

[36816 rows x 14 columns]

In [11]:
df_edges_sorted["Target_name"] = df_edges_sorted["Target"].map(df.set_index("Id")["name"])
df_edges_sorted

Source  Target      Type      Id  Label  Weight airline   aid    s  \
5558     3830    3682  Directed  276882    NaN      20      VS  5347  ORD   
4066     3682    3830  Directed  275354    NaN      19      VS  5347  ATL   
2685     3179    3885  Directed  273623    NaN      13      TG  4940  HKT   
5631     3830    3861  Directed  276955    NaN      13      US  5265  ORD   
4065     3682    3576  Directed  275353    NaN      12      VS  5347  ATL   
...       ...     ...       ...     ...    ...     ...     ...   ...  ...   
36811    3330    6337  Directed  338305    NaN       1      ZL  4178  TSV   
36812    3363    3339  Directed  338306    NaN       1      ZL  4178  WGA   
36813    6337    6289  Directed  338308    NaN       1      ZL  4178  WIN   
36814    6337    3330  Directed  338309    NaN       1      ZL  4178  WIN   
36815    6334    3341  Directed  338310    NaN       1      ZL  4178  WYA   

         d codeshare  stops                    equipment  \
5558   ATL         Y      0                  717 M88 319   
4066   ORD         Y      0      757 319 320 M90 M88 717   
2685   BKK         Y      0  333 744 330 AB6 773 747 320   
5631   MSY         Y      0                          E75   
4065   MIA         Y      0                      757 320   
...    ...       ...    ...                          ...   
36811  WIN       NaN      0                          SF3   
36812  MEL       NaN      0                          SF3   
36813  LRE       NaN      0                          SF3   
36814  TSV       NaN      0                          SF3   
36815  ADL       NaN      0                          SF3   

                                            Source_name  \
5558               Chicago O'Hare International Airport   
4066   Hartsfield Jackson Atlanta International Airport   
2685                       Phuket International Airport   
5631               Chicago O'Hare International Airport   
4065   Hartsfield Jackson Atlanta International Airport   
...                                                 ...   
36811                                Townsville Airport   
36812                          Wagga Wagga City Airport   
36813                                    Winton Airport   
36814                                    Winton Airport   
36815                                   Whyalla Airport   

                                             Target_name  
5558    Hartsfield Jackson Atlanta International Airport  
4066                Chicago O'Hare International Airport  
2685                                Suvarnabhumi Airport  
5631   Louis Armstrong New Orleans International Airport  
4065                         Miami International Airport  
...                                                  ...  
36811                                     Winton Airport  
36812                    Melbourne International Airport  
36813                                  Longreach Airport  
36814                                 Townsville Airport  
36815                     Adelaide International Airport  

[36816 rows x 15 columns]

In [12]:
df_top5 = df_edges_sorted.nlargest(30, "Weight")
df_top5

Source  Target      Type      Id  Label  Weight airline   aid    s  \
5558     3830    3682  Directed  276882    NaN      20      VS  5347  ORD   
4066     3682    3830  Directed  275354    NaN      19      VS  5347  ATL   
2685     3179    3885  Directed  273623    NaN      13      TG  4940  HKT   
5631     3830    3861  Directed  276955    NaN      13      US  5265  ORD   
4065     3682    3576  Directed  275353    NaN      12      VS  5347  ATL   
8250     3370    3386  Directed  279903    NaN      12      ZH  4611  CAN   
4969     3797     507  Directed  276275    NaN      12      VS  5347  JFK   
13877    3885    3077  Directed  288476    NaN      12      UL  4349  BKK   
4078     2179    2194  Directed  275367    NaN      12      WY  3778  AUH   
13910    3077    3885  Directed  288529    NaN      12      UL  4349  HKG   
4701    11051    2057  Directed  275997    NaN      12      WY  3778  DOH   
5337     3576    3682  Directed  276655    NaN      12      VS  5347  MIA   
5187      507    3797  Directed  276501    NaN      12      VS  5347  LHR   
5059     3484     507  Directed  276368    NaN      11      VS  5347  LAX   
8528     1005     907  Directed  280226    NaN      11      TK  4951  HRE   
3161     3861    3797  Directed  274248    NaN      11      VS  5347  MSY   
8422     3931    3885  Directed  280095    NaN      11      TG  4940  CNX   
7016     3751    3682  Directed  278520    NaN      11      WN  4547  DEN   
4948     3797    1382  Directed  276253    NaN      11      US  5265  JFK   
3151     3797    3861  Directed  274238    NaN      11      VS  5347  JFK   
4220     1382    3797  Directed  275511    NaN      11      US  5265  CDG   
4063     3682    3670  Directed  275351    NaN      11      VS  5347  ATL   
6971     3682    3751  Directed  278467    NaN      11      WN  4547  ATL   
8820     3469    3682  Directed  280589    NaN      11      WN  4547  SFO   
5302     2194    2179  Directed  276619    NaN      11      WY  3778  MCT   
5192      507    3484  Directed  276506    NaN      11      VS  5347  LHR   
7087     1165    1187  Directed  278595    NaN      11      WB  4292  KGL   
6195     3361    2006  Directed  277527    NaN      11      VA  5360  SYD   
626      3393    3379  Directed  271304    NaN      10      ZH  4611  CKG   
8559     3304    3275  Directed  280267    NaN      10      QZ  2857  KUL   

         d codeshare  stops                    equipment  \
5558   ATL         Y      0                  717 M88 319   
4066   ORD         Y      0      757 319 320 M90 M88 717   
2685   BKK         Y      0  333 744 330 AB6 773 747 320   
5631   MSY         Y      0                          E75   
4065   MIA         Y      0                      757 320   
8250   HGH         Y      0                      737 320   
4969   LHR         Y      0                  333 346 744   
13877  HKG         Y      0                          332   
4078   MCT         Y      0              ATR E75 738 73G   
13910  BKK         Y      0                          332   
4701   BAH         Y      0                          738   
5337   ATL         Y      0              M88 757 M90 73H   
5187   JFK         Y      0                  346 744 333   
5059   LHR         Y      0                  346 343 744   
8528   LUN         Y      0                          788   
3161   JFK         Y      0                          CR9   
8422   BKK         Y      0  AB6 747 734 320 333 773 777   
7016   ATL         Y      0                          73W   
4948   CDG         Y      0                      763 757   
3151   MSY         Y      0                          CR9   
4220   JFK         Y      0                      763 757   
4063   DFW         Y      0                      M88 738   
6971   DEN         Y      0                          73W   
8820   ATL         Y      0                          73W   
5302   AUH         Y      0              ATR E75 738 73G   
5192   LAX         Y      0                  346 343 744   

In [13]:
df_top5[["Weight","Source_name", "Target_name"]]

Weight                                        Source_name  \
5558       20               Chicago O'Hare International Airport   
4066       19   Hartsfield Jackson Atlanta International Airport   
2685       13                       Phuket International Airport   
5631       13               Chicago O'Hare International Airport   
4065       12   Hartsfield Jackson Atlanta International Airport   
8250       12             Guangzhou Baiyun International Airport   
4969       12               John F Kennedy International Airport   
13877      12                               Suvarnabhumi Airport   
4078       12                    Abu Dhabi International Airport   
13910      12                    Hong Kong International Airport   
4701       12                        Hamad International Airport   
5337       12                        Miami International Airport   
5187       12                            London Heathrow Airport   
5059       11                  Los Angeles International Airport   
8528       11        Robert Gabriel Mugabe International Airport   
3161       11  Louis Armstrong New Orleans International Airport   
8422       11                   Chiang Mai International Airport   
7016       11                       Denver International Airport   
4948       11               John F Kennedy International Airport   
3151       11               John F Kennedy International Airport   
4220       11            Charles de Gaulle International Airport   
4063       11   Hartsfield Jackson Atlanta International Airport   
6971       11   Hartsfield Jackson Atlanta International Airport   
8820       11                San Francisco International Airport   
5302       11                       Muscat International Airport   
5192       11                            London Heathrow Airport   
7087       11                       Kigali International Airport   
6195       11       Sydney Kingsford Smith International Airport   
626        10           Chongqing Jiangbei International Airport   
8559       10                 Kuala Lumpur International Airport   

                                             Target_name  
5558    Hartsfield Jackson Atlanta International Airport  
4066                Chicago O'Hare International Airport  
2685                                Suvarnabhumi Airport  
5631   Louis Armstrong New Orleans International Airport  
4065                         Miami International Airport  
8250             Hangzhou Xiaoshan International Airport  
4969                             London Heathrow Airport  
13877                    Hong Kong International Airport  
4078                        Muscat International Airport  
13910                               Suvarnabhumi Airport  
4701                       Bahrain International Airport  
5337    Hartsfield Jackson Atlanta International Airport  
5187                John F Kennedy International Airport  
5059                             London Heathrow Airport  
8528         Kenneth Kaunda International Airport Lusaka  
3161                John F Kennedy International Airport  
8422                                Suvarnabhumi Airport  
7016    Hartsfield Jackson Atlanta International Airport  
4948             Charles de Gaulle International Airport  
3151   Louis Armstrong New Orleans International Airport  
4220                John F Kennedy International Airport  
4063             Dallas Fort Worth International Airport  
6971                        Denver International Airport  
8820    Hartsfield Jackson Atlanta International Airport  
5302                     Abu Dhabi International Airport  
5192                   Los Angeles International Airport  
7087                       Entebbe International Airport  
6195                      Auckland International Airport  
626                 Xi'an Xianyang International Airport  
8559                Soekarno-Hatta International Airport

In [33]:
df1 = pd.read_csv("Flights_norm.csv")
df1

Id  Label                                         name          city  \
0         1    NaN                               Goroka Airport        Goroka   
1         2    NaN                               Madang Airport        Madang   
2         3    NaN                 Mount Hagen Kagamuga Airport   Mount Hagen   
3         4    NaN                               Nadzab Airport        Nadzab   
4         5    NaN  Port Moresby Jacksons International Airport  Port Moresby   
...     ...    ...                                          ...           ...   
3142  11258    NaN                      Saumlaki/Olilit Airport      Saumlaki   
3143  11436    NaN                           Tarko-Sale Airport    Tarko-Sale   
3144  11473    NaN                             Jinchuan Airport      Jinchuan   
3145  11498    NaN            Alashankou Bole (Bortala) airport          Bole   
3146  11922    NaN             Iwakuni Marine Corps Air Station       Iwakuni   

               country iata  icao   latitude   longitude   alt  ... indegree  \
0     Papua New Guinea  GKA  AYGA  -6.081690  145.391998  5282  ...        4   
1     Papua New Guinea  MAG  AYMD  -5.207080  145.789001    20  ...        7   
2     Papua New Guinea  HGU  AYMH  -5.826790  144.296005  5388  ...        9   
3     Papua New Guinea  LAE  AYNZ  -6.569803  146.725977   239  ...        9   
4     Papua New Guinea  POM  AYPY  -9.443380  147.220001   146  ...       32   
...                ...  ...   ...        ...         ...   ...  ...      ...   
3142         Indonesia  SXK  WAPI  -7.988610  131.306000   218  ...        1   
3143            Russia  TQL  USDS  64.930801   77.818100    82  ...        1   
3144             China  JIC  ZLJC  38.542222  102.348333  4740  ...        4   
3145             China  BPL  ZWAX  44.895000   82.300000  1253  ...        1   
3146             Japan  IWK  RJOI  34.143902  132.235992     7  ...        1   

     outdegree Degree  weighted indegree  weighted outdegree  Weighted Degree  \
0            4      8                  5                   5               10   
1            7     14                  8                   8               16   
2            8     17                 12                  10               22   
3            9     18                 11                  11               22   
4           32     64                 47                  48               95   
...        ...    ...                ...                 ...              ...   
3142         1      2                  1                   1                2   
3143         1      2                  1                   1                2   
3144         4      8                  6                   6               12   
3145         1      2                  1                   1                2   
3146         1      2                  1                   1                2   

      Eccentricity  closnesscentrality  harmonicclosnesscentrality  \
0                9            0.228983                    0.239638   
1                9            0.229033                    0.240114   
2                9            0.229050                    0.240273   
3                9            0.229067                    0.240432   
4                8            0.296876                    0.321212   
...            ...                 ...                         ...   
3142            11            0.164248                    0.168398   
3143            10            0.208745                    0.217677   
3144             9            0.231545                    0.245423   
3145             9            0.235092                    0.247446   
3146             9            0.257152                    0.269162   

      betweenesscentrality  
0                 0.000000  
1                 0.000001  
2                 0.000002  
3                 0.000002  
4                 0.015009  
...                    ...  
3142              0.000000  
3143              0.000000  


In [34]:
df1_sorted = df1.sort_values(by="betweenesscentrality", ascending=False)
df1_sorted.head(6)

Id  Label                                         name         city  \
626   1382    NaN      Charles de Gaulle International Airport        Paris   
1698  3484    NaN            Los Angeles International Airport  Los Angeles   
1006  2188    NaN                  Dubai International Airport        Dubai   
191    340    NaN                    Frankfurt am Main Airport    Frankfurt   
1843  3774    NaN  Ted Stevens Anchorage International Airport    Anchorage   
1628  3364    NaN        Beijing Capital International Airport      Beijing   

                   country iata  icao   latitude   longitude  alt  ...  \
626                 France  CDG  LFPG  49.012798    2.550000  392  ...   
1698         United States  LAX  KLAX  33.942501 -118.407997  125  ...   
1006  United Arab Emirates  DXB  OMDB  25.252800   55.364399   62  ...   
191                Germany  FRA  EDDF  50.033333    8.570556  364  ...   
1843         United States  ANC  PANC  61.174400 -149.996002  152  ...   
1628                 China  PEK  ZBAA  40.080101  116.584999  116  ...   

     indegree outdegree Degree  weighted indegree  weighted outdegree  \
626       232       237    469                516                 524   
1698      147       148    295                497                 489   
1006      179       185    364                348                 350   
191       238       239    477                493                 497   
1843       34        34     68                 59                  59   
1628      204       204    408                530                 531   

      Weighted Degree  Eccentricity  closnesscentrality  \
626              1040             7            0.411726   
1698              986             7            0.397825   
1006              698             7            0.404266   
191               990             7            0.414766   
1843              118             8            0.301919   
1628             1061             7            0.389356   

      harmonicclosnesscentrality  betweenesscentrality  
626                     0.464461              0.064090  
1698                    0.438457              0.062142  
1006                    0.451291              0.060087  
191                     0.466609              0.054225  
1843                    0.326639              0.052764  
1628                    0.441109              0.051084  

[6 rows x 23 columns]

In [29]:
df1_sorted.head(30)

Id  Label                                               name  \
626   1382    NaN            Charles de Gaulle International Airport   
1698  3484    NaN                  Los Angeles International Airport   
1006  2188    NaN                        Dubai International Airport   
191    340    NaN                          Frankfurt am Main Airport   
1843  3774    NaN        Ted Stevens Anchorage International Airport   
1628  3364    NaN              Beijing Capital International Airport   
1867  3830    NaN               Chicago O'Hare International Airport   
125    193    NaN            Lester B. Pearson International Airport   
282    580    NaN                         Amsterdam Airport Schiphol   
766   1701    NaN                      Atatürk International Airport   
1178  2564    NaN  Guarulhos - Governador André Franco Montoro In...   
255    507    NaN                            London Heathrow Airport   
1047  2279    NaN                       Narita International Airport   
1625  3361    NaN       Sydney Kingsford Smith International Airport   
1594  3316    NaN                           Singapore Changi Airport   
1597  3320    NaN                     Brisbane International Airport   
1737  3577    NaN               Seattle Tacoma International Airport   
1792  3682    NaN   Hartsfield Jackson Atlanta International Airport   
1996  4029    NaN                   Domodedovo International Airport   
1472  3077    NaN                    Hong Kong International Airport   
89     146    NaN  Montreal / Pierre Elliott Trudeau Internationa...   
1924  3930    NaN                      Incheon International Airport   
1852  3797    NaN               John F Kennedy International Airport   
1783  3670    NaN            Dallas Fort Worth International Airport   
1835  3751    NaN                       Denver International Airport   
1251  2709    NaN                    El Dorado International Airport   
381    813    NaN                     OR Tambo International Airport   
1727  3550    NaN       George Bush Intercontinental Houston Airport   
1900  3885    NaN                               Suvarnabhumi Airport   
576   1229    NaN               Adolfo Suárez Madrid–Barajas Airport   

                   city               country iata  icao   latitude  \
626               Paris                France  CDG  LFPG  49.012798   
1698        Los Angeles         United States  LAX  KLAX  33.942501   
1006              Dubai  United Arab Emirates  DXB  OMDB  25.252800   
191           Frankfurt               Germany  FRA  EDDF  50.033333   
1843          Anchorage         United States  ANC  PANC  61.174400   
1628            Beijing                 China  PEK  ZBAA  40.080101   
1867            Chicago         United States  ORD  KORD  41.978600   
125             Toronto                Canada  YYZ  CYYZ  43.677200   
282           Amsterdam           Netherlands  AMS  EHAM  52.308601   
766            Istanbul                Turkey  ISL  LTBA  40.976898   
1178          Sao Paulo                Brazil  GRU  SBGR -23.435556   
255              London        United Kingdom  LHR  EGLL  51.470600   
1047              Tokyo                 Japan  NRT  RJAA  35.764702   
1625             Sydney             Australia  SYD  YSSY -33.946098   
1594          Singapore             Singapore  SIN  WSSS   1.350190   
1597           Brisbane             Australia  BNE  YBBN -27.384199   
1737            Seattle         United States  SEA  KSEA  47.449001   
1792            Atlanta         United States  ATL  KATL  33.636700   
1996             Moscow                Russia  DME  UUDD  55.408798   
1472          Hong Kong             Hong Kong  HKG  VHHH  22.308901   
89             Montreal                Canada  YUL  CYUL  45.470600   
1924              Seoul           South Korea  ICN  RKSI  37.469101   
1852           New York         United States  JFK  KJFK  40.639801   
1783  Dallas-Fort Worth         United States  DFW  KDFW  32.896801   


In [28]:
df1_sorted2 = df1_sorted.sort_values(by="outdegree", ascending=False)
outdegree_desc = df1_sorted2.reset_index(drop=True)
outdegree_desc[outdegree_desc["name"] == "Ted Stevens Anchorage International Airport" ] # ranking outdegree = 279, BC = 5
outdegree_desc[outdegree_desc["name"] == "Guarulhos - Governador André Franco Montoro International Airport" ] # ranking outdegree = 74, BC = 11
outdegree_desc[outdegree_desc["name"] == "Sydney Kingsford Smith International Airport" ] # ranking outdegree = 85, BC = 14
outdegree_desc[outdegree_desc["name"] == "Brisbane International Airport" ] # ranking outdegree = 145, BC = 16

Id  Label                            name      city    country iata  \
145  3320    NaN  Brisbane International Airport  Brisbane  Australia  BNE   

     icao   latitude   longitude  alt  ... indegree outdegree Degree  \
145  YBBN -27.384199  153.117004   13  ...       55        59    114   

     weighted indegree  weighted outdegree  Weighted Degree  Eccentricity  \
145                143                 151              294             8   

     closnesscentrality  harmonicclosnesscentrality  betweenesscentrality  
145            0.329666                    0.359612               0.03193  

[1 rows x 23 columns]

In [35]:
df1_sorted2 = df1_sorted.sort_values(by="outdegree", ascending=False)

In [52]:
# Step 1: Reset index so we can get rankings
df1_ranked = df1_sorted.reset_index(drop=True)
df2_ranked = df1_sorted2.reset_index(drop=True)

# Step 2: Add rankings to each DataFrame
df1_ranked['bc_rank'] = df1_ranked.index +1
df2_ranked['outdegree_rank'] = df2_ranked.index +1

# Step 3: Merge on 'name' to align the data
merged = pd.merge(
    df1_ranked[['name', 'city', 'bc_rank']],
    df2_ranked[['name', 'outdegree_rank']],
    on='name'
)

# Step 4: Calculate the absolute difference in rankings
merged['rank_diff'] = (merged['bc_rank'] - merged['outdegree_rank']).abs()

# Step 5: Sort by difference
final = merged.sort_values(by='bc_rank', ascending=True)

# (Optional) Reset index for neatness
final = final.reset_index(drop=True)

df_final = pd.DataFrame(final)
# Show the top rows
df_final.head(50)

name               city  \
0             Charles de Gaulle International Airport              Paris   
1                   Los Angeles International Airport        Los Angeles   
2                         Dubai International Airport              Dubai   
3                           Frankfurt am Main Airport          Frankfurt   
4         Ted Stevens Anchorage International Airport          Anchorage   
5               Beijing Capital International Airport            Beijing   
6                Chicago O'Hare International Airport            Chicago   
7             Lester B. Pearson International Airport            Toronto   
8                          Amsterdam Airport Schiphol          Amsterdam   
9                       Atatürk International Airport           Istanbul   
10  Guarulhos - Governador André Franco Montoro In...          Sao Paulo   
11                            London Heathrow Airport             London   
12                       Narita International Airport              Tokyo   
13       Sydney Kingsford Smith International Airport             Sydney   
14                           Singapore Changi Airport          Singapore   
15                     Brisbane International Airport           Brisbane   
16               Seattle Tacoma International Airport            Seattle   
17   Hartsfield Jackson Atlanta International Airport            Atlanta   
18                   Domodedovo International Airport             Moscow   
19                    Hong Kong International Airport          Hong Kong   
20  Montreal / Pierre Elliott Trudeau Internationa...           Montreal   
21                      Incheon International Airport              Seoul   
22               John F Kennedy International Airport           New York   
23            Dallas Fort Worth International Airport  Dallas-Fort Worth   
24                       Denver International Airport             Denver   
25                    El Dorado International Airport             Bogota   
26                     OR Tambo International Airport       Johannesburg   
27       George Bush Intercontinental Houston Airport            Houston   
28                               Suvarnabhumi Airport            Bangkok   
29               Adolfo Suárez Madrid–Barajas Airport             Madrid   
30              Shanghai Pudong International Airport           Shanghai   
31                 Ninoy Aquino International Airport             Manila   
32                        Miami International Airport              Miami   
33                          Stockholm-Arlanda Airport          Stockholm   
34                    Vancouver International Airport          Vancouver   
35                 Kuala Lumpur International Airport       Kuala Lumpur   
36                        Hamad International Airport               Doha   
37             Guangzhou Baiyun International Airport          Guangzhou   
38               Soekarno-Hatta International Airport            Jakarta   
39                 Sheremetyevo International Airport             Moscow   
40               Newark Liberty International Airport             Newark   
41                 Tokyo Haneda International Airport              Tokyo   
42                                     Munich Airport             Munich   
43                        Faa'a International Airport            Papeete   
44                Indira Gandhi International Airport              Delhi   
45  Comodoro Arturo Merino Benítez International A...           Santiago   
46                     Auckland International Airport           Auckland   
47     Licenciado Benito Juarez International Airport        Mexico City   
48                              Jorge Newbery Airpark       Buenos Aires   
49                         Copenhagen Kastrup Airport         Copenhagen   

    bc_rank  outdegree_rank  rank_diff  
0         1               2          1  
1         2              23         21  
2         3              11          8  
3       

In [53]:
df_final[df_final["name"].isin([
    "Ted Stevens Anchorage International Airport",
    "Guarulhos - Governador André Franco Montoro International Airport",
    "Sydney Kingsford Smith International Airport"
])]

name       city  bc_rank  \
4         Ted Stevens Anchorage International Airport  Anchorage        5   
10  Guarulhos - Governador André Franco Montoro In...  Sao Paulo       11   
13       Sydney Kingsford Smith International Airport     Sydney       14   

    outdegree_rank  rank_diff  
4              280        275  
10              74         63  
13              86         72

In [59]:
df.groupby("country").count()

Id  Label  name  city  iata  icao  latitude  longitude  \
country                                                                     
Afghanistan         4      0     4     4     4     4         4          4   
Albania             1      0     1     1     1     1         1          1   
Algeria            26      0    26    26    26    26        26         26   
American Samoa      1      0     1     1     1     1         1          1   
Angola             14      0    14    14    14    14        14         14   
...                ..    ...   ...   ...   ...   ...       ...        ...   
Wallis and Futuna   2      0     2     2     2     2         2          2   
Western Sahara      2      0     2     2     2     2         2          2   
Yemen               9      0     9     9     9     9         9          9   
Zambia              7      0     7     7     7     7         7          7   
Zimbabwe            3      0     3     3     3     3         3          3   

                   alt  tz  ...  indegree  outdegree  Degree  \
country                     ...                                
Afghanistan          4   4  ...         4          4       4   
Albania              1   1  ...         1          1       1   
Algeria             26  26  ...        26         26      26   
American Samoa       1   1  ...         1          1       1   
Angola              14  14  ...        14         14      14   
...                ...  ..  ...       ...        ...     ...   
Wallis and Futuna    2   2  ...         2          2       2   
Western Sahara       2   2  ...         2          2       2   
Yemen                9   9  ...         9          9       9   
Zambia               7   7  ...         7          7       7   
Zimbabwe             3   3  ...         3          3       3   

                   weighted indegree  weighted outdegree  Weighted Degree  \
country                                                                     
Afghanistan                        4                   4                4   
Albania                            1                   1                1   
Algeria                           26                  26               26   
American Samoa                     1                   1                1   
Angola                            14                  14               14   
...                              ...                 ...              ...   
Wallis and Futuna                  2                   2                2   
Western Sahara                     2                   2                2   
Yemen                              9                   9                9   
Zambia                             7                   7                7   
Zimbabwe                           3                   3                3   

                   Eccentricity  closnesscentrality  \
country                                               
Afghanistan                   4                   4   
Albania                       1                   1   
Algeria                      26                  26   
American Samoa                1                   1   
Angola                       14                  14   
...                         ...                 ...   
Wallis and Futuna             2                   2   
Western Sahara                2                   2   
Yemen                         9                   9   
Zambia                        7                   7   
Zimbabwe                      3                   3   

                   harmonicclosnesscentrality  betweenesscentrality  
country                                                              
Afghanistan                                 4                     4  
Albania                                     1                     1  
Algeria                                    26                    26  
American Samoa                              1                     1  
Angola                                     14 

In [3]:
df_edges = pd.read_csv("Flights_Edges_notMerged.csv")
df_edges.head()

Source  Target      Type      Id  Label  Weight airline  aid    s    d  \
0    2965    2990  Directed  270652    NaN       1      2B  410  AER  KZN   
1    2966    2990  Directed  270653    NaN       1      2B  410  ASF  KZN   
2    2966    2962  Directed  270654    NaN       1      2B  410  ASF  MRV   
3    2968    2990  Directed  270655    NaN       1      2B  410  CEK  KZN   
4    2968    4078  Directed  270656    NaN       1      2B  410  CEK  OVB   

  codeshare  stops equipment  
0       NaN      0       CR2  
1       NaN      0       CR2  
2       NaN      0       CR2  
3       NaN      0       CR2  
4       NaN      0       CR2

In [4]:
df_edges.groupby("airline").count()

Source  Target  Type   Id  Label  Weight  aid    s    d  codeshare  \
airline                                                                       
2B           40      40    40   40      0      40   40   40   40          0   
2G           19      19    19   19      0      19   19   19   19          0   
2I           18      18    18   18      0      18   18   18   18          0   
2J           22      22    22   22      0      22   22   22   22          0   
2K           12      12    12   12      0      12   12   12   12          0   
...         ...     ...   ...  ...    ...     ...  ...  ...  ...        ...   
ZH          797     797   797  797      0     797  797  797  797        400   
ZI           64      64    64   64      0      64   64   64   64          0   
ZK           53      53    53   53      0      53   53   53   53          0   
ZL           88      88    88   88      0      88   88   88   88          0   
ZM            4       4     4    4      0       4    4    4    4          0   

         stops  equipment  
airline                    
2B          40         40  
2G          19         19  
2I          18         18  
2J          22         22  
2K          12         12  
...        ...        ...  
ZH         797        797  
ZI          64         64  
ZK          53         53  
ZL          88         88  
ZM           4          4  

[562 rows x 12 columns]

In [5]:
df_nodes = pd.read_csv("Flights_Nodes_notMerged.csv")
df_nodes.head()

Id  Label                                         name          city  \
0   1    NaN                               Goroka Airport        Goroka   
1   2    NaN                               Madang Airport        Madang   
2   3    NaN                 Mount Hagen Kagamuga Airport   Mount Hagen   
3   4    NaN                               Nadzab Airport        Nadzab   
4   5    NaN  Port Moresby Jacksons International Airport  Port Moresby   

            country iata  icao  latitude   longitude   alt  ... indegree  \
0  Papua New Guinea  GKA  AYGA -6.081690  145.391998  5282  ...        5   
1  Papua New Guinea  MAG  AYMD -5.207080  145.789001    20  ...        8   
2  Papua New Guinea  HGU  AYMH -5.826790  144.296005  5388  ...       12   
3  Papua New Guinea  LAE  AYNZ -6.569803  146.725977   239  ...       11   
4  Papua New Guinea  POM  AYPY -9.443380  147.220001   146  ...       47   

  outdegree Degree  weighted indegree  weighted outdegree  Weighted Degree  \
0         5     10                  5                   5               10   
1         8     16                  8                   8               16   
2        10     22                 12                  10               22   
3        11     22                 11                  11               22   
4        48     95                 47                  48               95   

   Eccentricity  closnesscentrality  harmonicclosnesscentrality  \
0             9            0.228983                    0.239638   
1             9            0.229033                    0.240114   
2             9            0.229050                    0.240273   
3             9            0.229067                    0.240432   
4             8            0.296876                    0.321212   

   betweenesscentrality  
0              0.000000  
1              0.000001  
2              0.000001  
3              0.000001  
4              0.013955  

[5 rows x 23 columns]

In [8]:
df_nodes_china = df_nodes.loc[df_nodes["country"] == "China"]
df_nodes_china #173 airports

Id  Label                                    name         city  \
1628   3364    NaN   Beijing Capital International Airport      Beijing   
1629   3366    NaN                        Dongshan Airport       Hailar   
1630   3368    NaN    Tianjin Binhai International Airport      Tianjin   
1631   3369    NaN                    Taiyuan Wusu Airport      Taiyuan   
1632   3370    NaN  Guangzhou Baiyun International Airport    Guangzhou   
...     ...    ...                                     ...          ...   
3134  10160    NaN   Erenhot Saiwusu International Airport      Erenhot   
3136  10940    NaN            Zhangjiakou Ningyuan Airport  Zhangjiakou   
3137  10941    NaN                  Arxan Yi'ershi Airport        Arxan   
3144  11473    NaN                        Jinchuan Airport     Jinchuan   
3145  11498    NaN       Alashankou Bole (Bortala) airport         Bole   

     country iata  icao   latitude   longitude   alt  ... indegree outdegree  \
1628   China  PEK  ZBAA  40.080101  116.584999   116  ...      530       531   
1629   China  HLD  ZBLA  49.205002  119.824997  2169  ...       18        18   
1630   China  TSN  ZBTJ  39.124401  117.346001    10  ...      145       145   
1631   China  TYN  ZBYN  37.746899  112.627998  2575  ...      109       109   
1632   China  CAN  ZGGG  23.392401  113.299004    50  ...      330       333   
...      ...  ...   ...        ...         ...   ...  ...      ...       ...   
3134   China  ERL  ZBER  43.422500  112.096667  3301  ...        2         2   
3136   China  ZQZ  ZBZJ  40.738602  114.930000  2347  ...        2         2   
3137   China  YIE  ZBES  47.310600  119.911700  2925  ...        1         1   
3144   China  JIC  ZLJC  38.542222  102.348333  4740  ...        6         6   
3145   China  BPL  ZWAX  44.895000   82.300000  1253  ...        1         1   

     Degree  weighted indegree  weighted outdegree  Weighted Degree  \
1628   1061                530                 531             1061   
1629     36                 18                  18               36   
1630    290                145                 145              290   
1631    218                109                 109              218   
1632    663                330                 333              663   
...     ...                ...                 ...              ...   
3134      4                  2                   2                4   
3136      4                  2                   2                4   
3137      2                  1                   1                2   
3144     12                  6                   6               12   
3145      2                  1                   1                2   

      Eccentricity  closnesscentrality  harmonicclosnesscentrality  \
1628             7            0.389356                    0.441109   
1629             8            0.281496                    0.300333   
1630             8            0.292896                    0.320870   
1631             8            0.291918                    0.318009   
1632             8            0.363489                    0.407522   
...            ...                 ...                         ...   
3134             8            0.280442                    0.297764   
3136             9            0.220989                    0.231133   
3137             9            0.224330                    0.234625   
3144             9            0.231545                    0.245423   
3145             9            0.235092                    0.247446   

      betweenesscentrality  
1628              0.057146  
1629              0.000206  
1630              0.000107  
1631              0.000024  
1632              0.015153  
...                    ...  
3134              0.000000  
3136              0.000000  
3137              0.000000  
3144              0.000001  
3145              0.000000  

[173 rows x 23 columns]

In [13]:
# Get list of IATA codes for Chinese airports
china_airports = set(df_nodes_china['iata'])

# Filter edges where both source and destination are in China
china_domestic_flights = df_edges[
    (df_edges['s'].isin(china_airports)) &
    (df_edges['d'].isin(china_airports))
]

len(china_domestic_flights)

6851

In [17]:
df_nodes_merged_USA = df.loc[df["country"] == "United States"]
df_nodes_merged_USA

Id  Label                        name                city  \
1669  3411    NaN  Barter Island LRRS Airport       Barter Island   
1670  3414    NaN      Point Lay LRRS Airport           Point Lay   
1671  3415    NaN  Hilo International Airport                Hilo   
1672  3417    NaN             Bettles Airport             Bettles   
1673  3420    NaN          Fort Yukon Airport          Fort Yukon   
...    ...    ...                         ...                 ...   
3054  8200    NaN         Toksook Bay Airport         Toksook Bay   
3070  8314    NaN         Hilton Head Airport  Hilton Head Island   
3112  9543    NaN      Ogden Hinckley Airport               Ogden   
3113  9739    NaN            Napakiak Airport            Napakiak   
3114  9744    NaN           Napaskiak Airport           Napaskiak   

            country iata  icao   latitude   longitude   alt  ... indegree  \
1669  United States  BTI  PABA  70.134003 -143.582001     2  ...        2   
1670  United States  PIZ  PPIZ  69.732903 -163.005005    22  ...        1   
1671  United States  ITO  PHTO  19.721399 -155.048004    38  ...        3   
1672  United States  BTT  PABT  66.913902 -151.529007   647  ...        3   
1673  United States  FYU  PFYU  66.571503 -145.250000   433  ...        4   
...             ...  ...   ...        ...         ...   ...  ...      ...   
3054  United States  OOK  PAOO  60.541401 -165.087006    59  ...        2   
3070  United States  HHH  KHXD  32.224400  -80.697502    19  ...        2   
3112  United States  OGD  KOGD  41.195900 -112.012001  4473  ...        1   
3113  United States  WNA  PANA  60.690300 -161.979004    17  ...        1   
3114  United States  PKA  PAPK  60.702900 -161.778000    24  ...        1   

     outdegree Degree  weighted indegree  weighted outdegree  Weighted Degree  \
1669         2      4                  2                   2                4   
1670         1      2                  1                   1                2   
1671         3      6                  4                   4                8   
1672         3      6                  4                   4                8   
1673         4      8                  6                   6               12   
...        ...    ...                ...                 ...              ...   
3054         1      3                  3                   1                4   
3070         2      4                  4                   4                8   
3112         1      2                  1                   1                2   
3113         1      2                  1                   1                2   
3114         1      2                  1                   1                2   

      Eccentricity  closnesscentrality  harmonicclosnesscentrality  \
1669            10            0.188654                    0.194670   
1670            10            0.188598                    0.194389   
1671             8            0.285300                    0.299329   
1672             9            0.212697                    0.220058   
1673             9            0.212754                    0.220349   
...            ...                 ...                         ...   
3054            11            0.158801                    0.162515   
3070             8            0.259550                    0.274628   
3112            10            0.208524                    0.216741   
3113            11            0.158801                    0.162515   
3114            10            0.188756                    0.195067   

      betweenesscentrality  
1669              9.500000  
1670              0.000000  
1671              0.000000  
1672              1.500000  
1673          13002.830581  
...                    ...  
3054           6289.000000  
3070              0.000000  
3112              0.000000  
3113           3145.000000  
3114           3145.000000  

[525 rows x 23 columns]

In [19]:
len(df_nodes_merged_USA.loc[df_nodes_merged_USA["outdegree"] >= 50])

32

In [26]:
df_nodes_portugal = df_nodes.loc[df_nodes["country"] == "Portugal"]
df_nodes_brazil = df_nodes.loc[df_nodes["country"] == "Brazil"]

portugal_airports = set(df_nodes_portugal['iata'])
brazil_airports = set(df_nodes_brazil['iata'])

PTtoBR_flights = df_edges[
    (df_edges['s'].isin(portugal_airports)) &
    (df_edges['d'].isin(brazil_airports))
]

BRtoPT_flights = df_edges[
    (df_edges['s'].isin(brazil_airports)) &
    (df_edges['d'].isin(portugal_airports))
]

print(len(PTtoBR_flights)+len(BRtoPT_flights))

24


In [27]:
df_edges.head()

Source  Target      Type      Id  Label  Weight airline  aid    s    d  \
0    2965    2990  Directed  270652    NaN       1      2B  410  AER  KZN   
1    2966    2990  Directed  270653    NaN       1      2B  410  ASF  KZN   
2    2966    2962  Directed  270654    NaN       1      2B  410  ASF  MRV   
3    2968    2990  Directed  270655    NaN       1      2B  410  CEK  KZN   
4    2968    4078  Directed  270656    NaN       1      2B  410  CEK  OVB   

  codeshare  stops equipment  
0       NaN      0       CR2  
1       NaN      0       CR2  
2       NaN      0       CR2  
3       NaN      0       CR2  
4       NaN      0       CR2

In [31]:
df_airlines = pd.read_csv("airlines.csv")


In [35]:
import re
df_airlines.loc[df_airlines["Private flight"] == "Ryanair"]


1 Private flight  \N   -  N/A Unnamed: 5 Unnamed: 6  Y
4288  4296        Ryanair  \N  FR  RYR    RYANAIR    Ireland  Y